File: api/app/services/stock_transfer.py
import uuid

from sqlalchemy import select
from sqlalchemy.orm import Session, joinedload

from app.models import InventoryBalance, Location, SKU, StockTransfer, StockTransferLine
from app.services.events import append_event


def _require_warehouses(db: Session, org_id: uuid.UUID, from_id: uuid.UUID, to_id: uuid.UUID) -> tuple[Location, Location]:
    if from_id == to_id:
        raise ValueError("Source and destination must differ")
    a = db.get(Location, from_id)
    b = db.get(Location, to_id)
    if not a or not b or a.org_id != org_id or b.org_id != org_id:
        raise ValueError("Locations not found or org mismatch")
    if a.type != "warehouse" or b.type != "warehouse":
        raise ValueError("Both endpoints must be warehouse locations")
    return a, b


def create_transfer(
    db: Session,
    org_id: uuid.UUID,
    from_location_id: uuid.UUID,
    to_location_id: uuid.UUID,
    lines: list[tuple[uuid.UUID, float]],
    notes: str | None,
    created_by_user_id: uuid.UUID | None,
) -> StockTransfer:
    _require_warehouses(db, org_id, from_location_id, to_location_id)
    seen: set[uuid.UUID] = set()
    for sku_id, qty in lines:
        if sku_id in seen:
            raise ValueError("Duplicate SKU in transfer lines")
        seen.add(sku_id)
        if qty <= 0:
            raise ValueError("Quantities must be positive")
        if not db.get(SKU, sku_id):
            raise ValueError(f"SKU {sku_id} not found")

    t = StockTransfer(
        org_id=org_id,
        from_location_id=from_location_id,
        to_location_id=to_location_id,
        status="draft",
        notes=notes,
        created_by_user_id=created_by_user_id,
    )
    db.add(t)
    db.flush()
    for sku_id, qty in lines:
        db.add(
            StockTransferLine(
                transfer_id=t.id,
                sku_id=sku_id,
                qty_requested=qty,
                qty_shipped=0,
                qty_received=0,
            )
        )
    append_event(
        db,
        "TransferCreated",
        {"transfer_id": str(t.id), "from": str(from_location_id), "to": str(to_location_id)},
        org_id=org_id,
        actor_user_id=created_by_user_id,
    )
    return t


def ship_transfer(
    db: Session,
    transfer_id: uuid.UUID,
    org_id: uuid.UUID,
    actor_user_id: uuid.UUID | None,
) -> StockTransfer:
    t = db.execute(
        select(StockTransfer)
        .where(StockTransfer.id == transfer_id, StockTransfer.org_id == org_id)
        .options(joinedload(StockTransfer.lines))
    ).scalar_one_or_none()
    if not t:
        raise ValueError("Transfer not found")
    if t.status != "draft":
        raise ValueError("Only draft transfers can be shipped")
    if not t.lines:
        raise ValueError("Transfer has no lines")

    for line in t.lines:
        inv = db.execute(
            select(InventoryBalance).where(
                InventoryBalance.location_id == t.from_location_id,
                InventoryBalance.sku_id == line.sku_id,
            ).with_for_update()
        ).scalar_one_or_none()
        if not inv:
            raise ValueError(f"No inventory row for SKU at source warehouse (line {line.id})")
        available = inv.on_hand - inv.reserved
        need = line.qty_requested
        if available < need:
            raise ValueError(f"Insufficient stock for SKU {line.sku_id}: need {need}, available {available}")
        inv.on_hand -= need
        line.qty_shipped = need

    t.status = "in_transit"
    append_event(
        db,
        "TransferShipped",
        {"transfer_id": str(t.id)},
        org_id=org_id,
        actor_user_id=actor_user_id,
    )
    return t


def receive_transfer(
    db: Session,
    transfer_id: uuid.UUID,
    org_id: uuid.UUID,
    actor_user_id: uuid.UUID | None,
) -> StockTransfer:
    t = db.execute(
        select(StockTransfer)
        .where(StockTransfer.id == transfer_id, StockTransfer.org_id == org_id)
        .options(joinedload(StockTransfer.lines))
    ).scalar_one_or_none()
    if not t:
        raise ValueError("Transfer not found")
    if t.status != "in_transit":
        raise ValueError("Only in_transit transfers can be received")

    for line in t.lines:
        qty = line.qty_shipped
        if qty <= 0:
            continue
        inv = db.execute(
            select(InventoryBalance).where(
                InventoryBalance.location_id == t.to_location_id,
                InventoryBalance.sku_id == line.sku_id,
            ).with_for_update()
        ).scalar_one_or_none()
        if not inv:
            inv = InventoryBalance(
                location_id=t.to_location_id,
                sku_id=line.sku_id,
                on_hand=0,
                reserved=0,
            )
            db.add(inv)
            db.flush()
        inv.on_hand += qty
        line.qty_received = qty

    t.status = "received"
    append_event(
        db,
        "TransferReceived",
        {"transfer_id": str(t.id)},
        org_id=org_id,
        actor_user_id=actor_user_id,
    )
    return t


def cancel_transfer(
    db: Session,
    transfer_id: uuid.UUID,
    org_id: uuid.UUID,
    actor_user_id: uuid.UUID | None,
) -> StockTransfer:
    t = db.get(StockTransfer, transfer_id)
    if not t or t.org_id != org_id:
        raise ValueError("Transfer not found")
    if t.status not in ("draft",):
        raise ValueError("Only draft transfers can be cancelled")
    t.status = "cancelled"
    append_event(
        db,
        "TransferCancelled",
        {"transfer_id": str(t.id)},
        org_id=org_id,
        actor_user_id=actor_user_id,
    )
    return t

File: web/lib/server/stockTransfer.ts
import type { Prisma } from "@prisma/client";
import { appendEvent } from "@/lib/server/events";

async function requireWarehouses(
  tx: Prisma.TransactionClient,
  orgId: string,
  fromId: string,
  toId: string,
) {
  if (fromId === toId) throw new Error("Source and destination must differ");
  const a = await tx.location.findUnique({ where: { id: fromId } });
  const b = await tx.location.findUnique({ where: { id: toId } });
  if (!a || !b || a.orgId !== orgId || b.orgId !== orgId) {
    throw new Error("Locations not found or org mismatch");
  }
  if (a.type !== "warehouse" || b.type !== "warehouse") {
    throw new Error("Both endpoints must be warehouse locations");
  }
  return { a, b };
}

export async function createTransfer(
  tx: Prisma.TransactionClient,
  orgId: string,
  fromLocationId: string,
  toLocationId: string,
  lines: { skuId: string; qty: number }[],
  notes: string | null,
  createdByUserId: string | null,
) {
  await requireWarehouses(tx, orgId, fromLocationId, toLocationId);
  const seen = new Set<string>();
  for (const { skuId, qty } of lines) {
    if (seen.has(skuId)) throw new Error("Duplicate SKU in transfer lines");
    seen.add(skuId);
    if (qty <= 0) throw new Error("Quantities must be positive");
    const sku = await tx.sKU.findUnique({ where: { id: skuId } });
    if (!sku) throw new Error(`SKU ${skuId} not found`);
  }

  const t = await tx.stockTransfer.create({
    data: {
      orgId,
      fromLocationId,
      toLocationId,
      status: "draft",
      notes: notes ?? undefined,
      createdByUserId: createdByUserId ?? undefined,
      lines: {
        create: lines.map((l) => ({
          skuId: l.skuId,
          qtyRequested: l.qty,
          qtyShipped: 0,
          qtyReceived: 0,
        })),
      },
    },
    include: { lines: true },
  });

  await appendEvent(
    tx,
    "TransferCreated",
    {
      transfer_id: t.id,
      from: fromLocationId,
      to: toLocationId,
    },
    orgId,
    createdByUserId,
  );
  return t;
}

export async function shipTransfer(
  tx: Prisma.TransactionClient,
  transferId: string,
  orgId: string,
  actorUserId: string | null,
) {
  const t = await tx.stockTransfer.findFirst({
    where: { id: transferId, orgId },
    include: { lines: true },
  });
  if (!t) throw new Error("Transfer not found");
  if (t.status !== "draft") throw new Error("Only draft transfers can be shipped");
  if (t.lines.length === 0) throw new Error("Transfer has no lines");

  for (const line of t.lines) {
    const inv = await tx.inventoryBalance.findFirst({
      where: { locationId: t.fromLocationId, skuId: line.skuId },
    });
    if (!inv) {
      throw new Error(`No inventory row for SKU at source warehouse (line ${line.id})`);
    }
    const available = inv.onHand - inv.reserved;
    const need = line.qtyRequested;
    if (available < need) {
      throw new Error(`Insufficient stock for SKU ${line.skuId}: need ${need}, available ${available}`);
    }
    await tx.inventoryBalance.update({
      where: { id: inv.id },
      data: { onHand: inv.onHand - need },
    });
    await tx.stockTransferLine.update({
      where: { id: line.id },
      data: { qtyShipped: need },
    });
  }

  await tx.stockTransfer.update({
    where: { id: transferId },
    data: { status: "in_transit" },
  });
  await appendEvent(tx, "TransferShipped", { transfer_id: t.id }, orgId, actorUserId);
  return tx.stockTransfer.findFirstOrThrow({
    where: { id: transferId },
    include: { lines: true },
  });
}

export async function receiveTransfer(
  tx: Prisma.TransactionClient,
  transferId: string,
  orgId: string,
  actorUserId: string | null,
) {
  const t = await tx.stockTransfer.findFirst({
    where: { id: transferId, orgId },
    include: { lines: true },
  });
  if (!t) throw new Error("Transfer not found");
  if (t.status !== "in_transit") throw new Error("Only in_transit transfers can be received");

  for (const line of t.lines) {
    const qty = line.qtyShipped;
    if (qty <= 0) continue;
    let inv = await tx.inventoryBalance.findFirst({
      where: { locationId: t.toLocationId, skuId: line.skuId },
    });
    if (!inv) {
      inv = await tx.inventoryBalance.create({
        data: {
          locationId: t.toLocationId,
          skuId: line.skuId,
          onHand: 0,
          reserved: 0,
        },
      });
    }
    await tx.inventoryBalance.update({
      where: { id: inv.id },
      data: { onHand: inv.onHand + qty },
    });
    await tx.stockTransferLine.update({
      where: { id: line.id },
      data: { qtyReceived: qty },
    });
  }

  await tx.stockTransfer.update({
    where: { id: transferId },
    data: { status: "received" },
  });
  await appendEvent(tx, "TransferReceived", { transfer_id: t.id }, orgId, actorUserId);
  return tx.stockTransfer.findFirstOrThrow({
    where: { id: transferId },
    include: { lines: true },
  });
}

export async function cancelTransfer(
  tx: Prisma.TransactionClient,
  transferId: string,
  orgId: string,
  actorUserId: string | null,
) {
  const t = await tx.stockTransfer.findFirst({
    where: { id: transferId, orgId },
  });
  if (!t) throw new Error("Transfer not found");
  if (t.status !== "draft") throw new Error("Only draft transfers can be cancelled");
  await tx.stockTransfer.update({
    where: { id: transferId },
    data: { status: "cancelled" },
  });
  await appendEvent(tx, "TransferCancelled", { transfer_id: t.id }, orgId, actorUserId);
  return tx.stockTransfer.findFirstOrThrow({
    where: { id: transferId },
    include: { lines: true },
  });
}
